[FSDP] verify backward_prefetch works correctly with unit test

weifengpy opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #107058

issue resolved: https://github.com/pytorch/pytorch/pull/105984

context:
* CI did not catch the commit that breaks backward_prefetch https://github.com/pytorch/pytorch/pull/105006
* we had an action item to add unit test to prevent similar cases: https://github.com/pytorch/pytorch/pull/105984

what's included in this unit test
* monkey patch
torch.distributed.fsdp._runtime_utils._get_handle_to_prefetch and check which handles are prefetched

for backward_prefetch = BackwardPrefetch.BACKWARD_PRE
* state._exec_order_data.handles_post_forward_order equals forward order: encoder 0...5 -> decoder 0...5 -> root
* pre-backward hook order: root -> decoder 5...0 -> encoder 5...0
* prefetch order: decoder 5...0 -> encoder 5...0 -> None
  * when current_handle=encoder 0, _get_handle_to_prefetch returns None

for backward_prefetch = BackwardPrefetch.BACKWARD_POST
* state._exec_order_data.handles_post_forward_order equals forward order: encoder 0...5 -> decoder 0...5 -> root
* post-backward hook (AccumulateGrad) order: decoder 5, 4...0 -> encoder 5...0 -> root
* prefetch order: decoder 4...0 -> encoder 5...0 -> None -> None
  * 1st None: when current_handle=encoder 0, _get_handle_to_prefetch returns None
  * 2nd None: when current_handle=root, we get decoder 5 inside _get_handle_to_prefetch but is not needed. so returns None

weifengpy committed:
[FSDP] verify backward_prefetch works correctly with unit test

issue resolved: https://github.com/pytorch/pytorch/pull/105984

context:
* CI did not catch the commit that breaks backward_prefetch https://github.com/pytorch/pytorch/pull/105006
* we had an action item to add unit test to prevent similar cases: https://github.com/pytorch/pytorch/pull/105984

what's included in this unit test
* monkey patch
torch.distributed.fsdp._runtime_utils._get_handle_to_prefetch and get
total function call and total non-None handle
* BackwardPrefetch.BACKWARD_POST and BackwardPrefetch.BACKWARD_PRE:
expect total non-None handle > 0
* backward_prefetch = None: expect total function call = 0

[ghstack-poisoned]

weifengpy committed:
Update on "[FSDP] verify backward_prefetch works correctly with unit test"


issue resolved: https://github.com/pytorch/pytorch/pull/105984

context:
* CI did not catch the commit that breaks backward_prefetch https://github.com/pytorch/pytorch/pull/105006
* we had an action item to add unit test to prevent similar cases: https://github.com/pytorch/pytorch/pull/105984

what's included in this unit test
* monkey patch
torch.distributed.fsdp._runtime_utils._get_handle_to_prefetch and get
total function call and total non-None handle
* BackwardPrefetch.BACKWARD_POST and BackwardPrefetch.BACKWARD_PRE:
expect total non-None handle > 0
* backward_prefetch = None: expect total function call = 0

[ghstack-poisoned]

weifengpy committed:
Update on "[FSDP] verify backward_prefetch works correctly with unit test"


issue resolved: https://github.com/pytorch/pytorch/pull/105984

context:
* CI did not catch the commit that breaks backward_prefetch https://github.com/pytorch/pytorch/pull/105006
* we had an action item to add unit test to prevent similar cases: https://github.com/pytorch/pytorch/pull/105984

what's included in this unit test
* monkey patch
torch.distributed.fsdp._runtime_utils._get_handle_to_prefetch and get
total function call and total non-None handle
* BackwardPrefetch.BACKWARD_POST and BackwardPrefetch.BACKWARD_PRE:
expect total non-None handle > 0
* backward_prefetch = None: expect total function call = 0

[ghstack-poisoned]

weifengpy committed:
Update on "[FSDP] verify backward_prefetch works correctly with unit test"


issue resolved: https://github.com/pytorch/pytorch/pull/105984

context:
* CI did not catch the commit that breaks backward_prefetch https://github.com/pytorch/pytorch/pull/105006
* we had an action item to add unit test to prevent similar cases: https://github.com/pytorch/pytorch/pull/105984

what's included in this unit test
* monkey patch
torch.distributed.fsdp._runtime_utils._get_handle_to_prefetch and get
total function call and total non-None handle
* BackwardPrefetch.BACKWARD_POST and BackwardPrefetch.BACKWARD_PRE:
expect total non-None handle > 0
* backward_prefetch = None: expect total function call = 0

[ghstack-poisoned]

awgu commented on test/distributed/fsdp/test_fsdp_backward_prefetch.py:
```diff
@@ -0,0 +1,126 @@
+        non_none_handle_count = 0
+        func_call_count = 0
+
+        def patched_get_handle_to_prefetch(*args, **kwargs):
+            handle = orig_get_handle_to_prefetch(*args, **kwargs)
```
Sorry that I am only thinking about how to do this now.

I think it would be valuable to hard code that the sequence of prefetched handles is the expected one. This test only checks that _some_ handles were prefetched, not that they were reasonable handles.

Let me know if you need help figuring out how map modules to handles.

weifengpy replied:
To verify the sequence of prefetched handles, I added DECODER_PARAM_FQNS and ENCODER_PARAM_FQNS

weifengpy committed:
Update on "[FSDP] verify backward_prefetch works correctly with unit test"


issue resolved: https://github.com/pytorch/pytorch/pull/105984

context:
* CI did not catch the commit that breaks backward_prefetch https://github.com/pytorch/pytorch/pull/105006
* we had an action item to add unit test to prevent similar cases: https://github.com/pytorch/pytorch/pull/105984

what's included in this unit test
* monkey patch
torch.distributed.fsdp._runtime_utils._get_handle_to_prefetch and get
total function call and total non-None handle
* BackwardPrefetch.BACKWARD_POST and BackwardPrefetch.BACKWARD_PRE:
expect total non-None handle > 0
* backward_prefetch = None: expect total function call = 0

[ghstack-poisoned]

weifengpy committed:
Update on "[FSDP] verify backward_prefetch works correctly with unit test"


issue resolved: https://github.com/pytorch/pytorch/pull/105984

context:
* CI did not catch the commit that breaks backward_prefetch https://github.com/pytorch/pytorch/pull/105006
* we had an action item to add unit test to prevent similar cases: https://github.com/pytorch/pytorch/pull/105984

what's included in this unit test
* monkey patch
torch.distributed.fsdp._runtime_utils._get_handle_to_prefetch and get
total function call and total non-None handle
* BackwardPrefetch.BACKWARD_POST and BackwardPrefetch.BACKWARD_PRE:
expect params from prefetch handle to follow order DECODER_PARAM_FQNS, ENCODER_PARAM_FQNS
* backward_prefetch = None: expect None prefetch handle

[ghstack-poisoned]

awgu commented on test/distributed/fsdp/test_fsdp_backward_prefetch.py:
```diff
@@ -0,0 +1,202 @@
+        tgt = torch.randn((20, 1, 1024), device="cuda")
+
+        # monkey patch
+        flat_param_fqns_array = []
```
super nit: When I have a `List[List[str]]` I tend to prefix the name with `all_`. In this case, I would say `all_handle_fqns: List[List[str]] = []`, and then each element would be `handles_fqns: List[str]`.

Maybe `array` is not as popular terminology in Python unless using an explicit array data structure (e.g. numpy or the array library)?

awgu commented on test/distributed/fsdp/test_fsdp_backward_prefetch.py:
```diff
@@ -0,0 +1,202 @@
+
+                for array_index, fqns in enumerate(flat_param_fqns_array):
+                    if array_index >= 0 and array_index < encoder_begin_index:
+                        param_index = encoder_begin_index - 1 - array_index
```
`param_index` seems more like `layer_index`? This is the layer index in the `nn.ModuleList`?

awgu commented on test/distributed/fsdp/test_fsdp_backward_prefetch.py:
```diff
@@ -0,0 +1,202 @@
+                    self.assertEqual(len(flat_param_fqns_array), 0)
+                elif backward_prefetch == BackwardPrefetch.BACKWARD_PRE:
+                    encoder_begin_index = ENCODER_BEGIN_INDEX_FOR_PRE
+                    # +1 is for None handle
```
Is the `None` handle from the root module? If not, then which module? This might be good to include in the comment.

weifengpy replied:
for BACKWARD_PRE, None is returned when current_handle = encoder 0 and index becomes -1 for eod.handles_post_forward_order

for BACKWARD_POST, None is returned in two cases
* when current_handle=encoder 0 and index becomes -1 for eod.handles_post_forward_order
* when current_handle=root, handles_post_forward_order returns decoder 5 but not needed

I updated PR description and added comments with more detail

awgu commented on test/distributed/fsdp/test_fsdp_backward_prefetch.py:
```diff
@@ -0,0 +1,202 @@
+                loss = model(src, tgt).sum()
+                loss.backward()
+                optim.step()
+                if backward_prefetch is None:
+                    self.assertEqual(len(flat_param_fqns_array), 0)
```
nit: Should we just `continue` after this branch since there is nothing to check to be able to unify the logic below slightly?

awgu reviewed (commented):
Thanks for refactoring this unit test to check the FQNs!

My main concern is that it is not immediately obvious to me reading the unit test how `BACKWARD_POST` and `BACKWARD_PRE` differ in their prefetch sequences. It seems like `PRE` prefetches one additional decoder handle compared to `POST`, yet the two prefetch the same number of handles total. This seems to mean that `POST` has one extra `None` handle in the sequence? If you could make this part more clear, that would be great.

weifengpy committed:
Update on "[FSDP] verify backward_prefetch works correctly with unit test"


issue resolved: https://github.com/pytorch/pytorch/pull/105984

context:
* CI did not catch the commit that breaks backward_prefetch https://github.com/pytorch/pytorch/pull/105006
* we had an action item to add unit test to prevent similar cases: https://github.com/pytorch/pytorch/pull/105984

what's included in this unit test
* monkey patch
torch.distributed.fsdp._runtime_utils._get_handle_to_prefetch and get
total function call and total non-None handle
* BackwardPrefetch.BACKWARD_POST and BackwardPrefetch.BACKWARD_PRE:
expect params from prefetch handle to follow order DECODER_PARAM_FQNS, ENCODER_PARAM_FQNS
* backward_prefetch = None: expect None prefetch handle

[ghstack-poisoned]

weifengpy commented on test/distributed/fsdp/test_fsdp_backward_prefetch.py:
```diff
@@ -0,0 +1,202 @@
+
+
+@no_type_check
+def get_flat_param_fqns(state: _FSDPState, handle: FlatParamHandle) -> None:
```
refactored into Optional[List[str]] to cover List[str] and None

weifengpy commented on test/distributed/fsdp/test_fsdp_backward_prefetch.py:
```diff
@@ -0,0 +1,202 @@
+
+
+@no_type_check
+def get_flat_param_fqns(state: _FSDPState, handle: FlatParamHandle) -> None:
```
refactored into _common_utils.py to keep code dry
* get_handle_fqns_from_root
* _log_post_backward_hook

weifengpy commented on test/distributed/fsdp/test_fsdp_backward_prefetch.py:
```diff
@@ -0,0 +1,202 @@
+
+
+@no_type_check
+def get_flat_param_fqns(state: _FSDPState, handle: FlatParamHandle) -> None:
```
thanks for explaining. I walked though the code path to understand the explanation

awgu commented on torch/distributed/fsdp/_common_utils.py:
```diff
@@ -351,6 +352,32 @@ def return_fn(param_to_fqns):
+
+
+@no_type_check
+def get_handle_fqns_from_root(
```
small nit: It does not matter, but I wonder if we should just add a prefix `_` on this function name for consistency with the others.

awgu commented on test/distributed/fsdp/test_fsdp_backward_prefetch.py:
```diff
@@ -0,0 +1,219 @@
+                    )
+                elif backward_prefetch == BackwardPrefetch.BACKWARD_POST:
+                    # state._exec_order_data.handles_post_forward_order
+                    # equals forward order
```
small nit:
```suggestion
                    # equals forward order (same as BACKWARD_PRE)
```

awgu commented on test/distributed/fsdp/test_fsdp_backward_prefetch.py:
```diff
@@ -0,0 +1,219 @@
+                    # prefetch order
+                    #     decoder 4...0 -> encoder 5...0 -> None -> None
+                    # 1st None: when current_handle=encoder 0,
+                    #           handles_post_forward_order returns None
```
nit: should "handles_post_forward_order returns None" be something like "_get_handle_to_prefetch returns None"? (same above)

awgu reviewed (approved):
This is amazing! Thanks for addressing feedback so fast.

This really helps us understand the differences between `BACKWARD_PRE` and `BACKWARD_POST`.

weifengpy commented on test/distributed/fsdp/test_fsdp_backward_prefetch.py:
```diff
@@ -0,0 +1,219 @@
+                    # 1st None: when current_handle=encoder 0,
+                    #           handles_post_forward_order returns None
+                    # 2nd None: when current_handle=root,
+                    #           handles_post_forward_order returns decoder 5 but not needed
```
right. wonder if we have a FSDP benchmark like dynamo 14k github? For example, I can add a logging to print how frequent is this mis-prefetch

weifengpy committed:
Update on "[FSDP] verify backward_prefetch works correctly with unit test"


issue resolved: https://github.com/pytorch/pytorch/pull/105984

context:
* CI did not catch the commit that breaks backward_prefetch https://github.com/pytorch/pytorch/pull/105006
* we had an action item to add unit test to prevent similar cases: https://github.com/pytorch/pytorch/pull/105984

what's included in this unit test
* monkey patch
torch.distributed.fsdp._runtime_utils._get_handle_to_prefetch and check which handles are prefetched

for backward_prefetch = BackwardPrefetch.BACKWARD_PRE
* state._exec_order_data.handles_post_forward_order equals forward order: encoder 0...5 -> decoder 0...5 -> root
* pre-backward hook order: root -> decoder 5...0 -> encoder 5...0
* prefetch order: decoder 5...0 -> encoder 5...0 -> None
  * when current_handle=encoder 0, _get_handle_to_prefetch returns None

for backward_prefetch = BackwardPrefetch.BACKWARD_POST
* state._exec_order_data.handles_post_forward_order equals forward order: encoder 0...5 -> decoder 0...5 -> root
* post-backward hook (AccumulateGrad) order: decoder 5, 4...0 -> encoder 5...0 -> root
* prefetch order: decoder 4...0 -> encoder 5...0 -> None -> None
  * 1st None: when current_handle=encoder 0, _get_handle_to_prefetch returns None
  * 2nd None: when current_handle=root, we get decoder 5 inside _get_handle_to_prefetch but is not needed. so returns None

[ghstack-poisoned]

awgu commented:
Updates look good to me. Let us merge!

weifengpy commented:
@pytorchmergebot merge

awgu commented:
I suspect that the merge conflict is a simple one, probably due to the HSDP PRs I landed recently. Sorry about that @weifengpy !

weifengpy committed:
Update on "[FSDP] verify backward_prefetch works correctly with unit test"


issue resolved: https://github.com/pytorch/pytorch/pull/105984

context:
* CI did not catch the commit that breaks backward_prefetch https://github.com/pytorch/pytorch/pull/105006
* we had an action item to add unit test to prevent similar cases: https://github.com/pytorch/pytorch/pull/105984

what's included in this unit test
* monkey patch
torch.distributed.fsdp._runtime_utils._get_handle_to_prefetch and check which handles are prefetched

for backward_prefetch = BackwardPrefetch.BACKWARD_PRE
* state._exec_order_data.handles_post_forward_order equals forward order: encoder 0...5 -> decoder 0...5 -> root
* pre-backward hook order: root -> decoder 5...0 -> encoder 5...0
* prefetch order: decoder 5...0 -> encoder 5...0 -> None
  * when current_handle=encoder 0, _get_handle_to_prefetch returns None

for backward_prefetch = BackwardPrefetch.BACKWARD_POST
* state._exec_order_data.handles_post_forward_order equals forward order: encoder 0...5 -> decoder 0...5 -> root
* post-backward hook (AccumulateGrad) order: decoder 5, 4...0 -> encoder 5...0 -> root
* prefetch order: decoder 4...0 -> encoder 5...0 -> None -> None
  * 1st None: when current_handle=encoder 0, _get_handle_to_prefetch returns None
  * 2nd None: when current_handle=root, we get decoder 5 inside _get_handle_to_prefetch but is not needed. so returns None

[ghstack-poisoned]

weifengpy commented:
> I suspect that the merge conflict is a simple one, probably due to the HSDP PRs I landed recently. Sorry about that @weifengpy !

no worries. I just rebased and will try merging again

weifengpy commented:
@pytorchmergebot merge
